### ISSUE TITLE
wasm: some e2e external test cases dont raise the expected error

### ISSUE THREAD

**Issue by patrick-east** [open-policy-agent/opa#2954]
There are a number of test cases which are configured to return an error upon evaluation, but when run with the Wasm SDK they are not generating an error. After a quick check it appears that they pretty much all instead yield an undefined result rather than an error. We should standardize the behavior where possible.

The list of tests include:

[test/cases/testdata/arithmetic/test-arithmetic-0818.yaml](https://github.com/open-policy-agent/opa/blob/master/test/cases/testdata/arithmetic/test-arithmetic-0818.yaml)
[test/cases/testdata/array/test-array-0042.yaml](https://github.com/open-policy-agent/opa/blob/master/test/cases/testdata/array/test-array-0042.yaml)
[test/cases/testdata/array/test-array-0043.yaml](https://github.com/open-policy-agent/opa/blob/master/test/cases/testdata/array/test-array-0043.yaml)
[test/cases/testdata/bitsshiftleft/test-bitsshiftleft-0065.yaml](https://github.com/open-policy-agent/opa/blob/master/test/cases/testdata/bitsshiftleft/test-bitsshiftleft-0065.yaml)
[test/cases/testdata/bitsshiftright/test-bitsshiftright-0070.yaml](https://github.com/open-policy-agent/opa/blob/master/test/cases/testdata/bitsshiftright/test-bitsshiftright-0070.yaml)
[test/cases/testdata/casts/test-casts-0827.yaml](https://github.com/open-policy-agent/opa/blob/master/test/cases/testdata/casts/test-casts-0827.yaml)
[test/cases/testdata/elsekeyword/test-elsekeyword-1066.yaml](https://github.com/open-policy-agent/opa/blob/master/test/cases/testdata/elsekeyword/test-elsekeyword-1066.yaml)
[test/cases/testdata/jsonbuiltins/test-jsonbuiltins-0926.yaml](https://github.com/open-policy-agent/opa/blob/master/test/cases/testdata/jsonbuiltins/test-jsonbuiltins-0926.yaml)
[test/cases/testdata/netcidrcontains/test-netcidrcontains-0101.yaml](https://github.com/open-policy-agent/opa/blob/master/test/cases/testdata/netcidrcontains/test-netcidrcontains-0101.yaml)
[test/cases/testdata/netcidrcontains/test-netcidrcontains-0100.yaml](https://github.com/open-policy-agent/opa/blob/master/test/cases/testdata/netcidrcontains/test-netcidrcontains-0100.yaml)
[test/cases/testdata/netcidrintersects/test-netcidrintersects-0090.yaml](https://github.com/open-policy-agent/opa/blob/master/test/cases/testdata/netcidrintersects/test-netcidrintersects-0090.yaml)
[test/cases/testdata/netcidrintersects/test-netcidrintersects-0091.yaml](https://github.com/open-policy-agent/opa/blob/master/test/cases/testdata/netcidrintersects/test-netcidrintersects-0091.yaml)
[test/cases/testdata/numbersrange/test-numbersrange-0260.yaml](https://github.com/open-policy-agent/opa/blob/master/test/cases/testdata/numbersrange/test-numbersrange-0260.yaml)
[test/cases/testdata/numbersrange/test-numbersrange-0261.yaml](https://github.com/open-policy-agent/opa/blob/master/test/cases/testdata/numbersrange/test-numbersrange-0261.yaml)
[test/cases/testdata/objectfilter/test-objectfilter-0315.yaml](https://github.com/open-policy-agent/opa/blob/master/test/cases/testdata/objectfilter/test-objectfilter-0315.yaml)
[test/cases/testdata/objectfilter/test-objectfilter-0317.yaml](https://github.com/open-policy-agent/opa/blob/master/test/cases/testdata/objectfilter/test-objectfilter-0317.yaml)
[test/cases/testdata/objectfilter/test-objectfilter-0316.yaml](https://github.com/open-policy-agent/opa/blob/master/test/cases/testdata/objectfilter/test-objectfilter-0316.yaml)
[test/cases/testdata/objectfilter/test-objectfilter-0314.yaml](https://github.com/open-policy-agent/opa/blob/master/test/cases/testdata/objectfilter/test-objectfilter-0314.yaml)
[test/cases/testdata/objectfilter/test-objectfilter-0309.yaml](https://github.com/open-policy-agent/opa/blob/master/test/cases/testdata/objectfilter/test-objectfilter-0309.yaml)
[test/cases/testdata/objectfilter/test-objectfilter-0310.yaml](https://github.com/open-policy-agent/opa/blob/master/test/cases/testdata/objectfilter/test-objectfilter-0310.yaml)
[test/cases/testdata/objectfilter/test-objectfilter-0313.yaml](https://github.com/open-policy-agent/opa/blob/master/test/cases/testdata/objectfilter/test-objectfilter-0313.yaml)
[test/cases/testdata/objectfilter/test-objectfilter-0311.yaml](https://github.com/open-policy-agent/opa/blob/master/test/cases/testdata/objectfilter/test-objectfilter-0311.yaml)
[test/cases/testdata/objectfilter/test-objectfilter-0312.yaml](https://github.com/open-policy-agent/opa/blob/master/test/cases/testdata/objectfilter/test-objectfilter-0312.yaml)
[test/cases/testdata/objectunion/test-objectunion-0277.yaml](https://github.com/open-policy-agent/opa/blob/master/test/cases/testdata/objectunion/test-objectunion-0277.yaml)
[test/cases/testdata/objectunion/test-objectunion-0278.yaml](https://github.com/open-policy-agent/opa/blob/master/test/cases/testdata/objectunion/test-objectunion-0278.yaml)
[test/cases/testdata/objectremove/test-objectremove-0289.yaml](https://github.com/open-policy-agent/opa/blob/master/test/cases/testdata/objectremove/test-objectremove-0289.yaml)
[test/cases/testdata/objectremove/test-objectremove-0290.yaml](https://github.com/open-policy-agent/opa/blob/master/test/cases/testdata/objectremove/test-objectremove-0290.yaml)
[test/cases/testdata/objectremove/test-objectremove-0291.yaml](https://github.com/open-policy-agent/opa/blob/master/test/cases/testdata/objectremove/test-objectremove-0291.yaml)
[test/cases/testdata/objectremove/test-objectremove-0292.yaml](https://github.com/open-policy-agent/opa/blob/master/test/cases/testdata/objectremove/test-objectremove-0292.yaml)
[test/cases/testdata/objectremove/test-objectremove-0293.yaml](https://github.com/open-policy-agent/opa/blob/master/test/cases/testdata/objectremove/test-objectremove-0293.yaml)
[test/cases/testdata/objectremove/test-objectremove-0294.yaml](https://github.com/open-policy-agent/opa/blob/master/test/cases/testdata/objectremove/test-objectremove-0294.yaml)
[test/cases/testdata/objectremove/test-objectremove-0295.yaml](https://github.com/open-policy-agent/opa/blob/master/test/cases/testdata/objectremove/test-objectremove-0295.yaml)
[test/cases/testdata/objectremove/test-objectremove-0296.yaml](https://github.com/open-policy-agent/opa/blob/master/test/cases/testdata/objectremove/test-objectremove-0296.yaml)
[test/cases/testdata/objectremove/test-objectremove-0297.yaml](https://github.com/open-policy-agent/opa/blob/master/test/cases/testdata/objectremove/test-objectremove-0297.yaml)
[test/cases/testdata/regexmatch/test-regexmatch-0857.yaml](https://github.com/open-policy-agent/opa/blob/master/test/cases/testdata/regexmatch/test-regexmatch-0857.yaml)
[test/cases/testdata/replacen/test-replacen-bad-operands.yaml](https://github.com/open-policy-agent/opa/blob/master/test/cases/testdata/replacen/test-replacen-bad-operands.yaml)
[test/cases/testdata/replacen/test-replacen-bad-operands.yaml](https://github.com/open-policy-agent/opa/blob/master/test/cases/testdata/replacen/test-replacen-bad-operands.yaml)
[test/cases/testdata/replacen/test-replacen-bad-operands.yaml](https://github.com/open-policy-agent/opa/blob/master/test/cases/testdata/replacen/test-replacen-bad-operands.yaml)
[test/cases/testdata/strings/test-strings-0881.yaml](https://github.com/open-policy-agent/opa/blob/master/test/cases/testdata/strings/test-strings-0881.yaml)
[test/cases/testdata/strings/test-strings-0892.yaml](https://github.com/open-policy-agent/opa/blob/master/test/cases/testdata/strings/test-strings-0892.yaml)
[test/cases/testdata/strings/test-strings-0892.yaml](https://github.com/open-policy-agent/opa/blob/master/test/cases/testdata/strings/test-strings-0892.yaml)

**Comment by srenatus** [comment]
I think some of them (if not many) would be _correct_ to return `undefined` if compared against evaluating them without strict builtin errors.

```
$ opa eval 'substring("foo", -1, 0)'
{}
$ opa eval --strict-builtin-errors 'substring("foo", -1, 0)'
{
  "errors": [
    {
      "message": "substring: negative offset",
      "code": "eval_builtin_error",
      "location": {
        "file": "",
        "row": 1,
        "col": 1
      }
    }
  ]
}
```

The "external tests" yaml specs expect strict builtin errors. What do we want here? Since I believe the non-strict error approach to be what we're striving towards, the tests should perhaps be updated, and evaluated in non-strict mode for the non-WASM tests?

**Comment by tsandall** [comment]
Updating the tests to expect undefined and evaluate in non-strict mode makes the most sense. It feels a bit odd to remove a bunch of expectations around error conditions but I can't come up with a reason to keep them.

**Comment by srenatus** [comment]
Alternatively, we could test these indirectly, so that

1. the topdown implementation is tested with strict builtin errors;
2. the wasm implementation is tested against topdown running without strict builtin errors

It's a bit odd considering that some builtins are still provided via calling out to their topdown implementations, but not much odder than now 😉 

The missing piece of information here, I think, is what people use. With different versions deployed, I'd guess _strict builtin errors_ might still be the more common thing. If that is true, we should not remove the expectations just yet.

**Comment by tsandall** [comment]
Since v0.25 just went out it's definitely not the norm. I'm also fine with just waiting because it does feel a bit rushed.

Why don't we just extend the tests to specify expected output for both modes. The topdown runner can execute either or both and assert on each. The wasm runner, since wasm only supports non-strict, can assert on the one. Thoughts?

**Comment by srenatus** [comment]
> Why don't we just extend the tests to specify expected output for both modes.

Yep, that sounds like a good way forward. 🚅

**Comment by srenatus** [comment]
Having a closer look, these are the ones not addressed by changing the builtin-error mode:

file | stmt | topdown | wasm | resolution 
---|----|---|----|---
`arithmetic/test-arithmetic-0818.yaml`|`1.1 % 1`| `eval_builtin_error` |`0.1`| fix wasm arithmetics
`elsekeyword/test-elsekeyword-1066.yaml` | two else keywords, true/false | `eval_conflict_error` | `{"x": false}` | fix wasm logic

**Comment by srenatus** [comment]
With #3022 merged, this will be fully resolved when #3031 is done. There're a few missing pieces there, however, that I'm going to pick up this week.

**Comment by srenatus** [comment]
I'll close this. Error checking has been made consistent. The only remaining issue we have (AFAIK) with topdown vs WASM is the golang SDK's calls to jsonpatch exploding: #2949.
